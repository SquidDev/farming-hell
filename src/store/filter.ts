import { type Filters, type OwnedServant, Priority, type Target } from "../data";

export const defaultFilters: Filters = Object.freeze({
  maxed: true,
  priority: Object.freeze({
    [Priority.High]: true,
    [Priority.Medium]: true,
    [Priority.Low]: true,
    [Priority.Unsummoned]: true,
  }),
});

const isMaxed = ({ current, target }: Target): boolean => current !== undefined && target !== undefined && current === target;
const isMaxedLax = ({ current, target }: Target): boolean => current === target;

export const matchesFilter = (filters: Filters, servant: OwnedServant): boolean => {
  if (filters.priority[servant.priority] === false) return false;

  return filters.maxed || !isMaxed(servant.level) || !servant.skills.every(isMaxed) || !servant.appendSkills.every(isMaxedLax);
};
