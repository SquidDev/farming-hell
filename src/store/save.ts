import { runInAction, toJS } from "mobx";

import save from "../support/save";
import type { Store } from ".";
import { Filters, Id, OwnedServant, Target, makeId } from "../data";
import { defaultFilters } from "./filter";
import { qpId } from "../data/constants";

type LatestSchema = {
  version: 1,
  servants: Array<OwnedServant>,
  items?: Array<[Id<"item">, number]>,
  filters?: Filters,
};

type Schemas
  = Array<OwnedServant>
  | LatestSchema;

const load = (store: Store, data: Schemas): void => {
  let servants: Array<OwnedServant> = [];
  const items: Map<Id<"item">, number> = new Map();
  let filters: Filters = defaultFilters;

  if (data instanceof Array) {
    servants = data;
  } else if (data.version === 1) {
    servants = data.servants;
    if (data.items) {
      for (const [id, count] of data.items) items.set(id, count);
    }
    if (data.filters) filters = data.filters;
  } else {
    throw new Error("Failed to parse this data.");
  }

  // Convert our legacy QP item (id=5) to the current one (id=1).
  const oldQpId = makeId("item", 5);
  const qp = items.get(oldQpId);
  if (qp !== undefined) {
    items.set(qpId, qp);
    items.delete(oldQpId);
  }

  runInAction(() => {
    store.ownedServants = servants;
    store.ownedItems = items;
    store.filters = filters;
  });
};

export const loadFromLocalStorage = (store: Store): void => {
  const servantData = window.localStorage.getItem("servants");
  if (!servantData) return;

  load(store, JSON.parse(servantData) as Schemas);
};

export const saveToLocalStorage = (store: Store): void => {
  const data: LatestSchema = {
    version: 1,
    servants: toJS(store.ownedServants),
    items: [...store.ownedItems.entries()],
    filters: store.filters,
  };
  window.localStorage.setItem("servants", JSON.stringify(data));
};

export const loadFromJson = (store: Store, blob: Blob): Promise<void> => {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.addEventListener("load", () => {
      try {
        load(store, JSON.parse(reader.result as string) as Schemas);
      } catch (e) {
        reject(`Failed to read file (${e})`); // eslint-disable-line @typescript-eslint/restrict-template-expressions
        return;
      }

      resolve();
    });
    reader.addEventListener("error", () => reject("Failed to read file"));
    reader.readAsText(blob);
  });
};

export const pickAndLoadFromJson = (store: Store): void => {
  const fileInput: HTMLInputElement = document.createElement("input");
  fileInput.type = "file";
  fileInput.accept = "application/json";
  fileInput.addEventListener("input", () => {
    const file = fileInput.files?.[0];
    if (!file) return;

    loadFromJson(store, file).catch(e => console.error(e));
  });
  fileInput.click();
};

export const saveToJson = (store: Store): void => {
  const data: LatestSchema = {
    version: 1,
    servants: toJS(store.ownedServants),
    items: [...store.ownedItems.entries()],
  };
  save("fgo-planner.json", new Blob([JSON.stringify(data)], { type: "application/json" }));
};

export const saveToCsv = (store: Store): void => {
  const target = ({ current, target }: Target): string => `${typeof current === "number" ? current : ""},${typeof target === "number" ? target : ""}`;

  const contents = store.ownedServants.map(servant => {
    const details = store.servantLookup.get(servant.id)!;
    return (
      `${servant.id},${details.name},${target(servant.level)},${target(servant.ascension)},` +
      `${target(servant.skills[0])},${target(servant.skills[1])},${target(servant.skills[2])}`
    );
  }).join("\n");
  save("fgo-planner.csv", new Blob([contents], { type: "text/csv" }));
};
