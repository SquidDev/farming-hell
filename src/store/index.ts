import { autorun, configure, makeAutoObservable } from "mobx";
import { v4 as uuidV4 } from "uuid";

import { DataDump, Filters, GameEvent, GrailCosts, Id, IdMap, Item, OwnedServant, Priority, Servant, makeId } from "../data";
import { Requirements, addServantRequirements, newRequirements } from "../data/requirements";
import { loadFromLocalStorage, saveToLocalStorage } from "./save";
import { defaultFilters } from "./filter";
import { expId, qpId } from "../data/constants";


if (process.env.NODE_ENV === "development") {
  configure({
    enforceActions: "always",
    computedRequiresReaction: true,
    reactionRequiresObservable: true,
    observableRequiresReaction: true,
    disableErrorBoundaries: true
  });
}

const makeLookup = <T extends { id: Id<string> }>(objects: Array<T>): IdMap<T> => {
  const lookup: IdMap<T> = new Map();
  for (const object of objects) lookup.set(object.id, object);
  return lookup;
};


/**
 * The global data store of servants.
 *
 * This is powered by mobx because I can't be dealing with Redux. Imperative
 * mutations to the store a propagated globally throughout the app.
 *
 * The store holds two bits of information:
 *  - All servants currently in NA. This is not tracked by mobx (and is
 *    notionally immutable).
 *  - The list of currently owned servants.
 */
export class Store {
  public readonly servants: Array<Servant>;
  public readonly items: Array<Item>;
  public readonly servantLookup: IdMap<Servant>;
  public readonly itemLookup: IdMap<Item>;
  public readonly eventLookup: IdMap<GameEvent>;
  public readonly expGrowth: Array<number>;
  public readonly grailCosts: GrailCosts;

  public ownedServants: Array<OwnedServant> = [];
  public ownedItems: Map<Id<"item">, number> = new Map();
  public filters: Filters = defaultFilters;

  constructor({ servants, items, events, expGrowth, grailCosts }: DataDump) {
    this.servants = servants;
    this.items = items;
    this.servantLookup = makeLookup(servants);
    this.itemLookup = makeLookup(items);
    this.eventLookup = makeLookup(events);
    this.expGrowth = expGrowth;
    this.grailCosts = grailCosts;

    makeAutoObservable(this, {
      servants: false,
      items: false,
      itemLookup: false,
      servantLookup: false,
      eventLookup: false,
      expGrowth: false,
      grailCosts: false,
    });
  }

  /**
   * Get requirements for all or a subset of servants.
   * @param filter Check whether this servant's requirements should be included.
   * @returns The accumulated requirements.
   */
  public getRequirements(filter?: (servant: OwnedServant, details: Servant) => boolean): Requirements {
    const requirements = newRequirements();
    for (const servant of this.ownedServants) {
      const details = this.servantLookup.get(servant.id);
      if (!details || (filter && !filter(servant, details))) continue;
      addServantRequirements(requirements, this, servant, details);
    }

    return requirements;
  }

  /** Get all required items. */
  public get requirements(): Requirements {
    return this.getRequirements();
  }

  public get requirementItems(): Array<{ item: Item, getter: (r: Requirements) => number }> {
    const result: Array<{ item: Item, getter: (r: Requirements) => number }> = [
      { item: this.itemLookup.get(qpId)!, getter: r => r.qp },
      { item: this.itemLookup.get(expId)!, getter: r => r.expBonus },
    ];

    for (const item of this.items) {
      if (item.id === qpId || item.id === expId) continue; // These have custom getters so skip.
      result.push({ item, getter: r => r.items.get(item.id) ?? 0 });
    }

    return result;
  }
}

export const artoriaPendragon: Id<"servant"> = makeId("servant", 100100);

/** Construct a new owned servant. */
export const newServant = (id?: Id<"servant">): OwnedServant => ({
  uid: uuidV4(),
  id: id ?? artoriaPendragon,
  priority: Priority.Low,
  level: {},
  ascension: {},
  skills: [{}, {}, {}],
});

/**
 * Construct a store driven from local storage. This is automatically saved
 * when the servants change.
 */
export const localStore = (data: DataDump): Store => {
  const store = new Store(data);

  try {
    loadFromLocalStorage(store);
  } catch (e) {
    console.error("Cannot load from local storage", e);
  }

  autorun(() => {
    try {
      saveToLocalStorage(store);
    } catch (e) {
      console.error("Cannot save to local storage", e);
    }
  });

  return store;
};
