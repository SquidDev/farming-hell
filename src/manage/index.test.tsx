import { useField } from "formik";
import { createElement } from "react";
import { create as render } from "react-test-renderer";

import { ServantInput } from ".";
import type { OwnedServant } from "../data";
import { Store, newServant } from "../store";
import { fsStore } from "../test";

const newStoreAndServant = async (): Promise<[Store, OwnedServant]> => {
  const store = await fsStore();
  const servant = newServant();
  store.ownedServants.push(servant);
  return [store, servant];
};

// Mock out some of the really noisy stuff
jest.mock("./ServantSelect", () => ({
  __esModule: true,
  default: () => createElement("ServantSelect"),
}));

jest.mock("../support/input", () => {
  const module: typeof import("../support/input") = jest.requireActual("../support/input");
  return {
    __esModule: true,
    ...module,
    Input: ({ name, placeholder }: { name: string, placeholder?: string }) => {
      const [field, _meta, _helpers] = useField<string>(name);
      return createElement("Input", { value: field.value, placeholder });
    },
  };
});

jest.mock("../items/requirements", () => {
  return {
    __esModule: true,
    ShortServantRequirements: () => createElement("ShortServantRequirements"),
  };
});

it("renders the empty servant correctly", async () => {
  const [store, servant] = await newStoreAndServant();

  const tree = render(<ServantInput servant={servant} store={store} />);
  expect(tree.toJSON()).toMatchSnapshot();
});

describe("Ascension and levels", () => {
  it("infers a servant's ascension correctly", async () => {
    const [store, servant] = await newStoreAndServant();
    servant.level = { current: 1, target: 90 };

    const tree = render(<ServantInput servant={servant} store={store} />);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("supports mixed ascension and level", async () => {
    const [store, servant] = await newStoreAndServant();
    servant.level = { current: 1, target: 90 };
    servant.ascension.current = 4;

    const tree = render(<ServantInput servant={servant} store={store} />);
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
