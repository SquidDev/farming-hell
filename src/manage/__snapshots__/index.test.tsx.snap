// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ascension and levels infers a servant's ascension correctly 1`] = `
Array [
  <div
    className="flex flex-col items-center sm:flex-row sm:items-start flex-nowrap gap-2 mb-4"
  >
    <img
      alt="Artoria Pendragon"
      className="w-20 h-20 squircle"
      height="128"
      src="/03906eed4894cd570263319006aa7cb2.png"
      width="128"
    />
    <div
      className="w-full flex flex-col flex-grow justify-center"
    >
      <ServantSelect />
      <select
        name="priority"
        onBlur={[Function]}
        onChange={[Function]}
        value="Low"
      >
        <option
          value="High"
        >
          High
        </option>
        <option
          value="Medium"
        >
          Medium
        </option>
        <option
          value="Low"
        >
          Low
        </option>
        <option
          value="Unsummoned"
        >
          Unsummoned
        </option>
      </select>
    </div>
  </div>,
  <div
    className="flex flex-col flex-grow"
  >
    <div
      className="flex flex-row flex-wrap mt-2 gap-2 2xl:gap-1"
    >
      <div
        className="rounded-md"
      >
        <div
          className="flex flex-col p-2"
        >
          <span
            className="text-xl"
          >
             
            Level
             
          </span>
          <div
            className="flex"
          >
            <Input
              value="1"
            />
            <Input
              value="90"
            />
          </div>
        </div>
      </div>
      <div
        className="rounded-md"
      >
        <div
          className="flex flex-col p-2"
        >
          <span
            className="text-xl"
          >
             
            Ascension
             
          </span>
          <div
            className="flex"
          >
            <Input
              placeholder="0"
              value=""
            />
            <Input
              placeholder="4"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="flex flex-col mt-2"
    >
      <span
        className="text-xl px-2"
      >
         Skills 
      </span>
      <div
        className="flex flex-wrap gap-2 2xl:gap-1"
      >
        <div
          className="rounded-md"
        >
          <div
            className="flex flex-col p-2"
          >
            <span
              className="text-xl"
            >
               
               
            </span>
            <div
              className="flex"
            >
              <Input
                value=""
              />
              <Input
                value=""
              />
            </div>
          </div>
        </div>
        <div
          className="rounded-md"
        >
          <div
            className="flex flex-col p-2"
          >
            <span
              className="text-xl"
            >
               
               
            </span>
            <div
              className="flex"
            >
              <Input
                value=""
              />
              <Input
                value=""
              />
            </div>
          </div>
        </div>
        <div
          className="rounded-md"
        >
          <div
            className="flex flex-col p-2"
          >
            <span
              className="text-xl"
            >
               
               
            </span>
            <div
              className="flex"
            >
              <Input
                value=""
              />
              <Input
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="mt-2"
    >
      <ShortServantRequirements />
    </div>
  </div>,
  <button
    className="absolute top-1 left-1 text-2xl w-6 h-6 flex items-center justify-center rounded-full bg-white bg-opacity-80 hover:bg-opacity-100 text-gray-900 hover:text-gray-600"
    onClick={[Function]}
    type="button"
  >
    ×
  </button>,
]
`;

exports[`Ascension and levels supports mixed ascension and level 1`] = `
Array [
  <div
    className="flex flex-col items-center sm:flex-row sm:items-start flex-nowrap gap-2 mb-4"
  >
    <img
      alt="Artoria Pendragon"
      className="w-20 h-20 squircle"
      height="128"
      src="/a268c4a24b6d8f7e8709fb84498299c7.png"
      width="128"
    />
    <div
      className="w-full flex flex-col flex-grow justify-center"
    >
      <ServantSelect />
      <select
        name="priority"
        onBlur={[Function]}
        onChange={[Function]}
        value="Low"
      >
        <option
          value="High"
        >
          High
        </option>
        <option
          value="Medium"
        >
          Medium
        </option>
        <option
          value="Low"
        >
          Low
        </option>
        <option
          value="Unsummoned"
        >
          Unsummoned
        </option>
      </select>
    </div>
  </div>,
  <div
    className="flex flex-col flex-grow"
  >
    <div
      className="flex flex-row flex-wrap mt-2 gap-2 2xl:gap-1"
    >
      <div
        className="rounded-md"
      >
        <div
          className="flex flex-col p-2"
        >
          <span
            className="text-xl"
          >
             
            Level
             
          </span>
          <div
            className="flex"
          >
            <Input
              value="1"
            />
            <Input
              value="90"
            />
          </div>
        </div>
      </div>
      <div
        className="bg-green-200 rounded-md"
      >
        <div
          className="flex flex-col p-2"
        >
          <span
            className="text-xl"
          >
             
            Ascension
             
          </span>
          <div
            className="flex"
          >
            <Input
              placeholder="4"
              value="4"
            />
            <Input
              placeholder="4"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="flex flex-col mt-2"
    >
      <span
        className="text-xl px-2"
      >
         Skills 
      </span>
      <div
        className="flex flex-wrap gap-2 2xl:gap-1"
      >
        <div
          className="rounded-md"
        >
          <div
            className="flex flex-col p-2"
          >
            <span
              className="text-xl"
            >
               
               
            </span>
            <div
              className="flex"
            >
              <Input
                value=""
              />
              <Input
                value=""
              />
            </div>
          </div>
        </div>
        <div
          className="rounded-md"
        >
          <div
            className="flex flex-col p-2"
          >
            <span
              className="text-xl"
            >
               
               
            </span>
            <div
              className="flex"
            >
              <Input
                value=""
              />
              <Input
                value=""
              />
            </div>
          </div>
        </div>
        <div
          className="rounded-md"
        >
          <div
            className="flex flex-col p-2"
          >
            <span
              className="text-xl"
            >
               
               
            </span>
            <div
              className="flex"
            >
              <Input
                value=""
              />
              <Input
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="mt-2"
    >
      <ShortServantRequirements />
    </div>
  </div>,
  <button
    className="absolute top-1 left-1 text-2xl w-6 h-6 flex items-center justify-center rounded-full bg-white bg-opacity-80 hover:bg-opacity-100 text-gray-900 hover:text-gray-600"
    onClick={[Function]}
    type="button"
  >
    ×
  </button>,
]
`;

exports[`renders the empty servant correctly 1`] = `
Array [
  <div
    className="flex flex-col items-center sm:flex-row sm:items-start flex-nowrap gap-2 mb-4"
  >
    <img
      alt="Artoria Pendragon"
      className="w-20 h-20 squircle"
      height="128"
      src="/03906eed4894cd570263319006aa7cb2.png"
      width="128"
    />
    <div
      className="w-full flex flex-col flex-grow justify-center"
    >
      <ServantSelect />
      <select
        name="priority"
        onBlur={[Function]}
        onChange={[Function]}
        value="Low"
      >
        <option
          value="High"
        >
          High
        </option>
        <option
          value="Medium"
        >
          Medium
        </option>
        <option
          value="Low"
        >
          Low
        </option>
        <option
          value="Unsummoned"
        >
          Unsummoned
        </option>
      </select>
    </div>
  </div>,
  <div
    className="flex flex-col flex-grow"
  >
    <div
      className="flex flex-row flex-wrap mt-2 gap-2 2xl:gap-1"
    >
      <div
        className="rounded-md"
      >
        <div
          className="flex flex-col p-2"
        >
          <span
            className="text-xl"
          >
             
            Level
             
          </span>
          <div
            className="flex"
          >
            <Input
              value=""
            />
            <Input
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="rounded-md"
      >
        <div
          className="flex flex-col p-2"
        >
          <span
            className="text-xl"
          >
             
            Ascension
             
          </span>
          <div
            className="flex"
          >
            <Input
              value=""
            />
            <Input
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="flex flex-col mt-2"
    >
      <span
        className="text-xl px-2"
      >
         Skills 
      </span>
      <div
        className="flex flex-wrap gap-2 2xl:gap-1"
      >
        <div
          className="rounded-md"
        >
          <div
            className="flex flex-col p-2"
          >
            <span
              className="text-xl"
            >
               
               
            </span>
            <div
              className="flex"
            >
              <Input
                value=""
              />
              <Input
                value=""
              />
            </div>
          </div>
        </div>
        <div
          className="rounded-md"
        >
          <div
            className="flex flex-col p-2"
          >
            <span
              className="text-xl"
            >
               
               
            </span>
            <div
              className="flex"
            >
              <Input
                value=""
              />
              <Input
                value=""
              />
            </div>
          </div>
        </div>
        <div
          className="rounded-md"
        >
          <div
            className="flex flex-col p-2"
          >
            <span
              className="text-xl"
            >
               
               
            </span>
            <div
              className="flex"
            >
              <Input
                value=""
              />
              <Input
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="mt-2"
    >
      <ShortServantRequirements />
    </div>
  </div>,
  <button
    className="absolute top-1 left-1 text-2xl w-6 h-6 flex items-center justify-center rounded-full bg-white bg-opacity-80 hover:bg-opacity-100 text-gray-900 hover:text-gray-600"
    onClick={[Function]}
    type="button"
  >
    ×
  </button>,
]
`;
